packages/dds/tree/src/codec/codec.ts

```diff
@@ -0,0 +1,177 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert, bufferToString, IsoBuffer } from "@fluidframework/common-utils";
+import { fail, JsonCompatibleReadOnly } from "../util";
+
+/**
+ * @alpha
+ */
+export interface IEncoder<TDecoded, TEncoded> {
+	/**
+	 * Encodes `obj` into some encoded format. Typically paired with an {@link IDecoder}.
+	 */
+	encode(obj: TDecoded): TEncoded;
+}
+
+/**
+ * @alpha
+ */
+export interface IDecoder<TDecoded, TEncoded> {
+	/**
+	 * Decodes `obj` from some encoded format. Typically paired with an {@link IEncoder}.
+	 */
+	decode(obj: TEncoded): TDecoded;
+}
+
+/**
+ * @alpha
+ */
+export interface IJsonCodec<TDecoded, TEncoded = JsonCompatibleReadOnly>
+	extends IEncoder<TDecoded, TEncoded>,
+		IDecoder<TDecoded, TEncoded> {}
+
+/**
+ * @remarks - TODO: We might consider using DataView or some kind of writer instead of IsoBuffer.
+ * @alpha
+ */
+export interface IBinaryCodec<TDecoded>
+	extends IEncoder<TDecoded, IsoBuffer>,
+		IDecoder<TDecoded, IsoBuffer> {}
+
+/**
+ * Contains knowledge of how to encode some in-memory type into JSON and binary formats,
+ * as well as how to decode those representations.
+ *
+ * @remarks - Codecs are typically used in shared-tree to convert data into some persisted format.
+ * For this common use case, any format for encoding that was ever actually used needs to
+ * be supported for decoding in all future code versions.
+ *
+ * Using an {@link ICodecFamily} is the recommended strategy for managing this support, keeping in
+ * mind evolution of encodings over time.
+ *
+ * @alpha
+ */
+export interface IMultiFormatCodec<TDecoded, TJsonEncoded = JsonCompatibleReadOnly> {
+	json: IJsonCodec<TDecoded, TJsonEncoded>;
+	binary: IBinaryCodec<TDecoded>;
+}
+
+/**
+ * Represents a family of codecs that can be used to encode and decode data in different formats.
+ * The family is identified by a format version, which is typically used to select the codec to use.
+ *
+ * Separating codecs into families rather than having a single codec support multiple versions (i.e. currying
+ * the `formatVersion` parameter)
+ * allows avoiding some duplicate work at encode/decode time, since the vast majority of document usage will not
+ * involve mixed format versions.
+ * @alpha
+ */
+export interface ICodecFamily<TDecoded> {
+	/**
+	 * @returns - a codec that can be used to encode and decode data in the specified format.
+	 */
+	resolve(formatVersion: number): IMultiFormatCodec<TDecoded>;
```

Comment:
what does this do in the case where there is no codec? Should have a way to return an error or be documented how it errors in this case.

I feel like apps will want specific telemetry for this case to help diagnose phased rollout issues, and might also want non-fatal handling to suggest a user update (non fatal errors should be documented catacahble errors or return values)

Reply:
having to iterate through the result of getSupportedFormats would work, but seems like not a good pattern to encourage. EIther this type should provide map like semantics or not (in which case the user can copy the results of getSupportedFormats into its own map/set)

Reply:
since you require a bounded set of keys anyway, why not just replace this type with a map?

Reply:
> what does this do in the case where there is no codec? Should have a way to return an error or be documented how it errors in this case.

My plan here was to throw.
I think this should be fine for consumers--throw should result in a dataProcessingError, which should hit whatever configured alerts the application may have if volume is high enough. Agree that there should be enough information in telemetry to determine the problematic piece of configuration. There is some logic to log well-known properties of fluid errors to telemetry, which we could either leverage or emit events of our own (no strong opinion on which is better right now, and I don't think it should be documented here anyway).

In terms of user experience, we can't really expect anything other than container close. Maybe we want to add some `ContainerErrorType` which specifically suggests upgrading, but current state isn't too bad for most hosts (container close with dataProcessingError causes an error banner and suggests the user refreshes) and I don't think whatever surfacing we do to the user should be very close to this API.

> since you require a bounded set of keys anyway, why not just replace this type with a map?

No reason why a map couldn't work. I went with this for a couple minor reasons:

1. suggested "throw" pattern above doesn't line up with map's `.get` semantics. so you'd want to do something like define a free function getter and use that everywhere.
2. I could imagine map being slightly less ergonomic for certain composition scenarios--say you had a data type written by a tree of codecs (much like editManager does) and wanted to update a single case in a way which warranted a new format (e.g. `valueField`'s encoding). You'd need to bump formatVersions for each of the other types of fields in the discriminated union and add a separate map entry for the same codec. This structure gives you the opportunity to express things like "formats 2 through 6 are the same" directly.